### ISSUE TITLE
Semantics of `registered_domain` property for private domains

### ISSUE THREAD

**Issue by tuler** [john-kurkowski/tldextract#138]
Suppose the following url: `tuler.github.io`
`github.io` is a private domain in the PSL.

When parsed with `include_psl_private_domains=True` we get `subdomain=''`, `domain=tuler`, `suffix=github.io`.

The `registered_domain` property just joins `domain` and `suffix`, giving me `tuler.github.io`, but IMHO it still should be `github.io`, as this is the domain registered with the registrar, and can be found in a whois query.

One problem to implement this is that when a URL is parsed, we can't know if the parsed domain is a private domain or a ICANN domain, because this is not kept internally when the PSL is read.

Any thoughts?

**Comment by john-kurkowski** [comment]
(Note to self, if we need to track public vs. private at runtime, #66 is a requirement.)

**Comment by john-kurkowski** [comment]
Yeah, I bet most will associate it with registrar registration, as you have.

In my mind, tldextract has been consistent, working as designed, via a more abstract interpretation of "registered." Excluding private domains, GitHub registered [github.io](//github.io) with a registrar, who controlled the domain. Including private domains, GitHub user tuler "registered" [tuler.github.io](//tuler.github.io) with GitHub, who controlled the domain.

I have no strong evidence if my interpretation is broadly useful. It was for a very specific case, when I originally wrote this lib. Or maybe both interpretations are useful.

**Comment by tuler** [comment]
I see your point.

Nonetheless, keeping runtime information regarding each domain from the PSL can be useful to handle this appropriately by the application. Something like a `is_private` method, or a `is_private` flag added to the `ExtractResult`.

**Comment by john-kurkowski** [comment]
Yes, at the very least we should do #66 and expose `is_private`.

I'd then consider a new `registered_domain`-like that was constant in the face of not/private. Just needs a new name.

Renaming today's `registered_domain` is also a possibility, but then we're burdened with backwards compat and legacy association with today's wording.

**Comment by john-kurkowski** [comment]
The PR for #66 currently tracks the source of an extraction, whether the official public suffix list, the private domains in the public suffix list, or user-provided extra suffixes. We haven't figured out how to expose that yet. It's tricky, since it's a `namedtuple`.

**Comment by gregtws** [comment]
Perhaps adding new APIs using terminology reflected by the PSL and libraries like Google Guavas' [InternetDomainName](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/net/InternetDomainName.html) would help?

It uses terms like public suffix and registry suffix to differentiate (lots of verbose docs so its super clear what you're getting).   So for `www.example.za.com` the public suffix is `za.com` and registry suffix is `com`.  Similarly it has explicit accessors like `topPrivateDomain` (1 level above the public suffix, so `example.za.com`) and `topDomainUnderRegistrySuffix` (1 level above the registry suffix, so `za.com`).

**Comment by john-kurkowski** [comment]
Good suggestion. With #66 long fixed, the new, constant-in-the-face-of-not/private property can be added. And maybe other properties. I'll draw on Guava as naming inspiration, and clarify `registered_domain`s docs.

**Comment by john-kurkowski** [comment]
@tuler @gregtws in #344,  I better documented the existing property, added a clearer alias `top_domain_under_public_suffix`, and added `top_domain_under_registry_suffix` that is constant between `include_psl_private_domains={False,True}`. Let me know what you think.
